### PR TITLE
[GEOT-7411] App-schema performance improvement in setting attribute values

### PR DESCRIPTION
**Pull request by groldan**
[GEOT-7411](https://osgeo-org.atlassian.net/browse/GEOT-7411) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

While building a complex attribute with a large sequence of children, XPath.set(...) incurs in a huge performance penalty due to excessive collections allocation and the inherent inefficiency of `ComplexAttributeImpl.getProperties(Name)` and
`ComplexAttributeImpl.setValue(Collection<Property>)`.

This is a pure refactoring patch on app-schema's `XPath` class, and the addition of the following non-api methods to gt-main's complex features implementation classes:
- `ComplexAttributeImpl.find(Name):Optional<Property>`
- `ComplexAttributeImpl.findLast(Name):Optional<Property>`
- `ComplexAttributeImpl.find(Predicate<Property>):Optional<Property>`
- `ComplexAttributeImpl.findAll(Name):Stream<Property>`
- `ComplexAttributeImpl.addValue(Property value):void`
- `PropertyImpl.getUserData(Object):Object`

With this in place, avoiding millions of object allocations, a use case where a WFS GetFeature request for a single complex feature contains over a hundred thousand related properties passed to take 2 seconds instead of 17 minutes, with `XPath.set(...)` taking only 7% of the whole request processing, where it used to take 95%.

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geotools.org/latest/developer/procedures/contribution_license.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] Avoid [Java 9+ split packages](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed).
- [ ] All the build checks are green ([see automated QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html)).

For core and extension modules:

- [ ] New unit tests have been added covering the changes.
- [ ] [Documentation](https://github.com/geotools/geotools/tree/main/docs) has been updated (if change is visible to end users).
- [x] There is an issue in [GeoTools Jira](https://osgeo-org.atlassian.net/projects/GEOT) (except for changes not visible to end users). 
- [x] Commit message(s) must be in the form ``[GEOT-XYZW] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] The commit targets a single objective (if multiple focuses cannot be avoided, each one is in its own commit, and has a separate ticket describing it).

### CODE REVIEW

**Comment by danduk82** [comment]
@groldan great thanks!
one stupid question: are you sure about the Geotools Jira ticket that you reference? It's one about logging and WMTS

**Comment by groldan** [comment]
Hi @aaime and/or @NielsCharlier, this is ready for review, and passing all downstream GeoServer tests .
Chees,

**Comment by aaime** [comment]
On my side the one that knows app-schema the best is @nmco , he's currently enjoying vacation but I'd love to hear from him once he comes back.

**Comment by NielsCharlier** [comment]
I am not maintainer, I don't work on app-schema much any more, but I have in the past.

The code itself looks good to me. App-schema does have a lot of test coverage (fortunately), have you run all the tests in geoserver, including the online tests?

File: modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java [hunk 1]
```diff
@@ -35,13 +35,11 @@ public abstract class PropertyImpl implements Property {
     /** descriptor of the property */
     protected PropertyDescriptor descriptor;
     /** user data */
-    protected final Map<Object, Object> userData;
+    private Map<Object, Object> _userData;
```
Comment (nmco) [inline modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java:38]: Why the change on the access modifier and variable name? This class is not final and hence could have been extended by third-party extensions, this will break the backward compatibility.

**Review by nmco — CHANGES_REQUESTED** [review]

**Comment by nmco** [comment]
Hi @groldan, the refactored methods are indeed quite complex; I reviewed your refactoring and only add the comment above.

That said, here we need to rely on the existing integration and online tests to validate that these core changes didn't break anything, can you confirm that the following tests were executed and are green:
- All the integration and online tests in GeoTools.
- [All the App-Schema integration tests in GeoServer](https://github.com/geoserver/geoserver/tree/main/src/extension/app-schema/app-schema-test).
- At least the online [GeoPackage](https://github.com/geoserver/geoserver/tree/main/src/extension/app-schema/app-schema-geopkg-test) and [Postgis](https://github.com/geoserver/geoserver/tree/main/src/extension/app-schema/app-schema-postgis-test) online tests for App-Schema in GeoServer.

Thank you!

File: modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java (hunk 1 above)
Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java:38]: Good point, also, not following the adopted [coding convention](https://google.github.io/styleguide/javaguide.html#s5.1-identifier-names):

> In Google Style, special prefixes or suffixes are not used. For example, these names are not Google Style: name_, mName, s_name and kName.

Comment (NielsCharlier) [reply modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java:38]: I think the reason for this change is that the field is now being lazily initialized. Therefore, subclasses should no longer use it directly but use the getter instead, otherwise they could get NPE. In that sense backwards compatibility for third-party subclasses is broken either way.

If we want the lazy initialization, this is the better option. Third parties will get a compilation fault when they upgrade geotools and are forced to fix it, in the other case they may not notice and get an NPE in production.

Comment (aaime) [reply modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java:38]: Agreed on the lazy initialization changing the API anyways. Could not find users of the protected field outside of PropertyImpl. The coding convention bit still stands.

Comment (groldan) [reply modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java:38]: Good point. Indeed it's due to the lazy initialization, which is beneficial because not all instances will have metadata. Re the variable name I can change it to something else that both adheres to the coding conventions and would result in a compile error of an eventual third party extension.

**Comment by groldan** [comment]
Thanks @nmco,
I didn't find any documentation on how to run the online tests, what needs to be prepared upfront, etc.
Do you know if there's something somewhere?

**Comment by nmco** [comment]
_I know it's complex. To me, it's now less complex than it was before, I kind of hate 200+ lines of code methods and prefer to refactor it so that you can understand what a method does without further digging into details unless you want to._ 

True 😄.

@groldan the documentation si available here:
https://docs.geoserver.org/latest/en/developer/programming-guide/app-schema/index.html

Let me know fi you have trouble running them.

**Comment by nmco** [comment]
Hi @groldan did you have any trouble running the online tests?

**Comment by groldan** [comment]
Hi @nmco, I've been working on something else and just now I could come back to this.
Let you know how it goes, thanks

**Comment by groldan** [comment]
@nmco it's hard to know whether the online tests will pass because I can't make them pass on main without this PR applied.
Also:
- The `geoserver-main-app-schema-online` build job is disabled:  https://build.geoserver.org/view/geoserver/job/geoserver-main-app-schema-online/
- The github action build for Oracle online tests hasn't run in the last 7 months and all previous runs also failed: https://github.com/geoserver/geoserver/actions/workflows/oracle_online.yml

So I wonder if someone runs them at all?

File: modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java (hunk 1 above)
Comment (groldan) [reply modules/library/main/src/main/java/org/geotools/feature/PropertyImpl.java]: Fixed: https://github.com/geotools/geotools/pull/4410/commits/45de0776540f133dae2073e270c22abc860f3756#diff-964e5fab63efd6e389520b1ecaf6b5bd87d769993a614c5568842437e90de9b4L37-L85

**Comment by jodygarnett** [comment]
@groldan because this was not merged before the refactor it has ended up with conflicts; can you rebase please.

**Comment by aaime** [comment]
Trying to figure out what the situation is, testing-wise (I'm normally not involved in it).

I can confirm tests are not passing. Here are some details:

The app-schema integration tests (not specific to any database, or at least, they should not be) have 30 tests failing out of 300 (the app-schema-online-test profile must be enabled, otherwise the failing tests are not being run):

```
ERROR] Failures: 
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnNestedAttribute:58->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnNestedAttribute2:71->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnNestedAttributeWithMaxNumber:82->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnRootAttribute:46->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithAndNestedFilterOnSameTypes:118->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithComplexPropertyORSimpleProperty:130->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[4]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithComplexPropertyORSimplePropertyWithPagination:155->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[4]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithFilterOnNestedAttribute:95->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithFilterOnNestedAttribute2:107->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithGeomComplexFilter:178->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithGeomComplexFilterManyAND:202->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithGeomComplexFilterWithPagination:190->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithMultipleAND:166->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithSimplePropertyANDComplexProperty:142->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedGMLTest.testGetMappedNumberMatchedWithIdFilter:212->assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnNestedAttribute:58->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnNestedAttribute2:71->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnNestedAttributeWithMaxNumber:82->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureHitsCountWithFilterOnRootAttribute:46->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithAndNestedFilterOnSameTypes:118->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithComplexPropertyORSimpleProperty:130->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[4]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithComplexPropertyORSimplePropertyWithPagination:155->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[4]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithFilterOnNestedAttribute:95->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithFilterOnNestedAttribute2:107->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithGeomComplexFilter:178->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithGeomComplexFilterManyAND:202->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithGeomComplexFilterWithPagination:190->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[3]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithMultipleAND:166->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetFeatureNumberMatchedWithSimplePropertyANDComplexProperty:142->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR]   GetFeatureNumberMatchedWithoutPrimaryKeyGMLTest>GetFeatureNumberMatchedGMLTest.testGetMappedNumberMatchedWithIdFilter:212->GetFeatureNumberMatchedGMLTest.assertNumberMathcedAndNumberReturned:217->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[1]> but was:<[0]>
[ERROR] Errors: 
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithAndNestedFilterOnDifferentTypes:25->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithAndNestedFilterOnSameTypes:37->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithComplexPropertyORSimpleProperty:49->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithComplexPropertyORSimplePropertyWithPagination:74->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithGeomComplexFilter:97->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithGeomComplexFilterManyAND:120->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithGeomComplexFilterWithPagination:109->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithMultipleAND:85->assertNumberMatchedAndNumberReturned:126 » JSON
[ERROR]   GetFeatureNumberMatchedGeoJSONTest.testGetFeatureNumberMatchedWithSimplePropertyANDComplexProperty:61->assertNumberMatchedAndNumberReturned:126 » JSON
[INFO] 
[ERROR] Tests run: 366, Failures: 30, Errors: 9, Skipped: 121
```

Once the ``postgis.properties`` file is configured, the postgis tests do run, and the following fail:

```
[ERROR] Failures: 
[ERROR]   FeatureChainingWfsTest.testEncodeFeatureMember:1868->AbstractAppSchemaTestSupport.assertXpathCount:340 expected:<1> but was:<0>
[ERROR]   PagingTest.testGetFeatureWithCSVFormat:900 expected:<text/csv[]> but was:<text/csv[;charset=UTF-8]>
[ERROR]   SRSWfsTest.testGetFeatureContent:148->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[-1.2 52.5 -1.2 52.6 -1.1 52.6 -1.1 52.5 -1.2 52.5]> but was:<[52.5 -1.2 52.6 -1.2 52.6 -1.1 52.5 -1.1 52.5 -1.2]>
[ERROR]   SRSWfsTest.testReproject:378->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[-1.2 52.5 -1.2 52.6 -1.1 52.6 -1.1 52.5 -1.2 52.5]> but was:<[52.50000004 -1.2 52.60000004 -1.2 52.60000004 -1.1 52.50000004 -1.1 52.50000004 -1.2]>
[INFO] 
[ERROR] Tests run: 318, Failures: 4, Errors: 0, Skipped: 0
```

The geopackage tests are odd, they have a sample configuration file in ``$HOME/.geoserver/.geopackage.properties.example`` that is a copy of the PostgreSQL one and well... they could be run offline really, just check that the current platform is supported by the GeoPackage library. In any case, fixing the sample file, they do run and have a subset of the PostGIS failures:

```
[ERROR] Failures: 
[ERROR]   FeatureChainingWfsTest.testEncodeFeatureMember:1868->AbstractAppSchemaTestSupport.assertXpathCount:340 expected:<1> but was:<0>
[ERROR]   PagingTest.testGetFeatureWithCSVFormat:900 expected:<text/csv[]> but was:<text/csv[;charset=UTF-8]>
```

I did not run the Oracle ones.

Funny enough, some modules are outside of the normal build and have mis-formatted pom and java files, as well. IMHO they should be part of the build, ensuring that at least they compile, and have tests be skipped if the online testing profile is not enabled.

**Comment by groldan** [comment]
little update, time allotted to work on this next January

**Comment by aaime** [comment]
@groldan did you manage to check if these changes are introducing new test failures, or not?

**Comment by groldan** [comment]
I can confirm we have the exact same failures for the postgis online tests as in the `main` branch:

```
[ERROR] Failures: 
[ERROR]   FeatureChainingWfsTest.testEncodeFeatureMember:1868->AbstractAppSchemaTestSupport.assertXpathCount:340 expected:<1> but was:<0>
[ERROR]   PagingTest.testGetFeatureWithCSVFormat:900 expected:<text/csv[]> but was:<text/csv[;charset=UTF-8]>
[ERROR]   SRSWfsTest.testGetFeatureContent:148->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[-1.2 52.5 -1.2 52.6 -1.1 52.6 -1.1 52.5 -1.2 52.5]> but was:<[52.5 -1.2 52.6 -1.2 52.6 -1.1 52.5 -1.1 52.5 -1.2]>
[ERROR]   SRSWfsTest.testReproject:378->AbstractAppSchemaTestSupport.assertXpathEvaluatesTo:329 expected:<[-1.2 52.5 -1.2 52.6 -1.1 52.6 -1.1 52.5 -1.2 52.5]> but was:<[52.50000004 -1.2 52.60000004 -1.2 52.60000004 -1.1 52.50000004 -1.1 52.50000004 -1.2]>
[INFO] 
[ERROR] Tests run: 318, Failures: 4, Errors: 0, Skipped: 0
```

I'll try to fix the easy ones in `main` after the PSC meeting today

**Comment by groldan** [comment]
Managed to fix the three easier pre-existing test failures, see https://github.com/geoserver/geoserver/pull/7424
And can confirm the same exact one remaining is the only failure with and without this patch.

**Comment by aaime** [comment]
Merged!

**Comment by aaime** [comment]
Wondering about backports? 30.x maybe?
